src/daemon/gnet/pool.go

```diff
@@ -115,523 +117,568 @@ type Connection struct {
+		messageState: state,
+		memChannel:   make(chan func()),
+	}
+
 	return pool
 }
 
-// Creates a new Connection around a net.Conn.  Trying to make a connection
-// to an address that is already connected will panic.
-func (self *ConnectionPool) NewConnection(conn net.Conn) *Connection {
-	a := conn.RemoteAddr().String()
-	if self.Addresses[a] != nil {
-		log.Panicf("Already connected to %s", a)
-	}
-	self.connId++
-	c := NewConnection(self, self.connId, conn,
-		self.Config.ConnectionWriteQueueSize)
-	self.connectionQueue <- c
-	return c
-}
+// Run starts the connection pool
+func (pool *ConnectionPool) Run() {
+	pool.Quit = make(chan struct{})
+	go func() {
+		for {
+			select {
+			case memActionFunc := <-pool.memChannel:
+				// this goroutine will handle all member variable's writing and reading actions.
+				memActionFunc()
+			case <-pool.Quit:
+				return
+			}
+		}
+	}()
 
-// Accepts incoming connections and sends them off for processing.
-// This function blocks, you will want to run it in a goroutine.
-func (self *ConnectionPool) AcceptConnections() {
-	logger.Debug("Accepting connections...")
-	if self.listener == nil {
-		log.Panic("Not listening, call StartListen first")
+	// start the connection accept loop
+	addr := fmt.Sprintf("%s:%v", pool.Config.Address, pool.Config.Port)
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Panic(err)
 	}
-	defer func() {
-		// If Accept fails for a reason other than StopListen closing listener,
-		// this could block indefinitely
-		self.acceptLock <- true
+
+	pool.listener = ln
+
+	go func() {
+		for {
+			logger.Info("Listening for connections...")
+			conn, err := ln.Accept()
+			if err != nil {
+				// logger.Error("%v", err)
+
+				// When Accept() returns with a non-nill error, we check the quit
+				// channel to see if we should continue or quit. If quit, then we quit.
+				// Otherwise we continue
+				select {
+				case <-pool.Quit:
+					return
+				default:
+					// without the default case the select will block.
+				}
+
+				continue
+			}
+
+			go pool.handleConnection(conn, false)
+		}
 	}()
-	// Grab the lock
-	<-self.acceptLock
-	// If we raced with StopListen and it beat us, listener will be nil and
-	// the intent was for us to stop.
-	if self.listener == nil {
-		return
-	}
-	for {
-		// this blocks until connection or error
-		conn, err := self.listener.Accept()
-		if err != nil {
-			logger.Debug("No longer accepting connections")
-			break
+}
+
+// Shutdown gracefully shutdown the connection pool
+func (pool *ConnectionPool) Shutdown() {
+	close(pool.Quit)
+	pool.listener.Close()
+	pool.listener = nil
+	pool.pool = make(map[int]*Connection)
+	pool.addresses = make(map[string]*Connection)
+}
+
+// NewConnection creates a new Connection around a net.Conn.  Trying to make a connection
+// to an address that is already connected will panic.
+func (pool *ConnectionPool) NewConnection(conn net.Conn) (*Connection, error) {
```

Comment:
It's possible to use
```go
func (pool *ConnectionPool) NewConnection(conn net.Conn) (nc *Connection, err error) {
    // blah
   return // nc, err
}
```
instead of 
```go
var nc *Connection
var err error
// blah
return nc, err
```